include/alpaka/block/shared/st/detail/BlockSharedMemStMemberImpl.hpp

```diff
@@ -0,0 +1,190 @@
+            auto operator=(BlockSharedMemStMemberImpl const&) -> BlockSharedMemStMemberImpl& = delete;
+            //-----------------------------------------------------------------------------
+            auto operator=(BlockSharedMemStMemberImpl&&) -> BlockSharedMemStMemberImpl& = delete;
+            //-----------------------------------------------------------------------------
+            /*virtual*/ ~BlockSharedMemStMemberImpl() = default;
+
+            template<typename T>
+            void alloc(uint32_t id) const
+            {
+                // Get meta data for the variable allocated by this function.
+                MetaData& lastMetaData = getLatestVar<MetaData>();
+
+                // Allocate variable
+                m_allocdBytes = varChunkSize<T>(m_allocdBytes);
+
+                // Add empty meta data chunk to the end
+                m_allocdBytes = varChunkSize<MetaData>(m_allocdBytes);
+                MetaData& meta = getLatestVar<MetaData>();
+                meta = MetaData{};
+
+                // Update meta data with id and offset for the allocated variable.
+                lastMetaData.id = id;
+                lastMetaData.offset = m_allocdBytes - metaDataSize;
+
+
+#if(defined ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST) && (!defined NDEBUG)
+                ALPAKA_ASSERT(m_allocdBytes <= m_capacity);
+#endif
+            }
+
+#if BOOST_COMP_GNUC
+#    pragma GCC diagnostic push
+#    pragma GCC diagnostic ignored                                                                                    \
+        "-Wcast-align" // "cast from 'unsigned char*' to 'unsigned int*' increases required alignment of target type"
+#endif
+
+            //! Give the pointer to an exiting variable
+            //!
+            //! @tparam T type of the variable
+            //! @param id unique id of the variable
+            //! @return nullptr if variable with id not exists
+            template<typename T>
+            T* getVar(uint32_t id) const
+            {
+                // Get first meta data header.
+                MetaData* metaDataPtr = reinterpret_cast<MetaData*>(m_mem);
+                // Offset in bytes to the next meta data header behind the variable.
+                uint32_t off = 0;
+
+                // Last meta data header offset is zero
+                while(metaDataPtr->offset != 0)
+                {
+                    off = metaDataPtr->offset;
+                    if(metaDataPtr->id == id)
+                        break;
+                    // next meta data header
+                    metaDataPtr = reinterpret_cast<MetaData*>(m_mem + off);
+                }
+
+                // Last header, variable with id not found.
+                if(metaDataPtr->offset == 0)
+                {
+                    return nullptr;
+                }
+
+                return reinterpret_cast<T*>(&m_mem[off - sizeof(T)]);
```

Comment:
Will this not break if the alignment of `T` is smaller than that of `MetaData`, i.e. if there are padding bytes before the address `metaDataPtr`?

Reply:
Note: @jkelling's comment was from before the refactoring.
In the refactored implementation `off` is pointing to the end of the variable, therefore `off - sizeof(T)` will point to an aligned offset.